### PR TITLE
Less strict size check in scatter

### LINKED ISSUE #63265: torch.scatter_: support index.size(d) > src.size(d)
## 🚀 Feature
Teach [torch.Tensor.scatter_](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html) to handle `index.size(d) > src.size(d)`.

## Motivation

Currently, `torch.Tensor.scatter_` requires `index.size(d) <= src.size(d)` for all dimensions `d`, _unless_ `src` is float-valued. This constraint seems artificial.

```
import math
import torch

device = 'cuda'  # fails if device == 'cpu' (cf. #61854)

dim = 1
shape_src = 3, 5
shape_idx = 3, 7

index = torch.empty(shape_idx, dtype=torch.long, device=device).random_(0, shape_src[dim])

# succeeds
output = torch.zeros(shape_src, device=device)
output.scatter_(dim, index, 1.0, reduce='add')
print(output)

# desired usage
# output = torch.zeros(shape_src, device=device)
# values = torch.rand(shape_src, device=device)
# output.scatter_(dim, index, values, reduce='add')   # <-- failure
# print(output)
```

## Alternatives

An inefficient workaround is to chunk the index tensor:
```
output = torch.zeros(shape_src, device=device)
values = torch.ones(shape_src, device=device)
chunks = math.ceil(index.shape[dim] / output.shape[dim])
for index_chunk in torch.chunk(index, chunks, dim):
    output.scatter_(dim, index_chunk, values, reduce='add')
print(output)
```

### PR DESCRIPTION
**Pull request by redmvc**
Fixes #63265.

This relaxes the constraint in scatterlike functions where all dimensions of the index tensor must be smaller than or equal to the size of the same dimensions of the source tensor. This is done by creating a separate pathway specific for this use case and then rebuilding the implied position of a given entry in the index tensor then constructing the corresponding position in the source tensor and recalculating the offset for it. Both CUDA and CPU are implemented and should work identically.

I added tests to check that everything is still working, but there were a couple of failures when the tests were run on `device = 'cuda'` and `dtype` was a float of some sort. In those cases, there seems to be a small (usually sub 1%) fraction of the tested entries that have a small (always sub 1.5%) relative difference between actual and expected values. I ran the tests locally without using the "src smaller than index" flag and still ran into those differences, though, so I'm not sure they were introduced by my changes.

I'm a first-time contributor to pyTorch so this probably has substantial issues or bits where I did not follow best practices. Nevertheless, it seems to work, and at least for generic use cases I tested does not seem much slower, though it is _somewhat_ slower as it requires some extra computation (O(ndim)). For normal use cases (where index has no dimension larger than the same dimension in source) the original code is used instead, so the effect of this change on existing applications should be minimal.

### CODE REVIEW

**Comment by redmagnos** [comment]
So @mikaylagawarecki I'm not sure how the process works and whether you're waiting for all checks to pass before taking a first look (I'm new at this so don't know what the general procedures are ^^') but as far as I can tell there is one specific check that's failing - the tests I added for the feature. However, the failures are very small (as I mentioned in the original PR message), and at least when I ran the tests locally on a branch that doesn't have my changes, the existing tests still failed at similar rates.

Which does kind of leave me confused, though, as to how other branches/PRs pass all tests? Do they not always check the scatter/gather ops and only do this when those are changed, or something? Or is there something I should do on my end?

File: test/test_scatter_gather_ops.py [hunk 1]
```diff
@@ -98,7 +98,11 @@ def _test_scatter_base(self, fn, *, device, dtype, is_scalar, reduction,
         if is_scalar:
             src = random.random()
         else:
-            src_size = [random.randint(1, 5) + s for s in idx_size]
+            sign = [1, 1, 1]
+            if src_smaller_than_index:
+                while(min(sign) > 0):
+                    sign = [random.choice([-1, 1]) if s > 1 else 1 for s in idx_size]
```
Comment (mikaylagawarecki) [inline test/test_scatter_gather_ops.py:104]: does this mean we don't generate cases where src is smaller than index on dim 0?

**Comment by mikaylagawarecki** [comment]
@redmagnos Thank you for contributing as well as your patience on the review! Regarding the tolerance issues on CUDA, that is likely because scatter/scatter_add/scatter_reduce use atomics on CUDA which are nondeterministic. It is interesting that you are seeing failures on these tests before the PR though as these tests should be running in CI. Let me try to reproduce this locally.

**Comment by nikitaved** [comment]
Thank you, @redmagnos , for this work! I will definitely have a look at it. Scatter is nondeterministic, but you can make it deterministic by using a source with all-equal elements!

File: test/test_scatter_gather_ops.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def test_gather_backward_with_empty_index_tensor(self, device, dtype, sparse_gra
         self.assertEqual(grad, expected_grad, atol=0, rtol=0)
 
     def _test_scatter_base(self, fn, *, device, dtype, is_scalar, reduction,
-                           unique_indices=True, include_self=True):
+                           unique_indices=True, include_self=True, src_smaller_than_index=False):
```
Comment (mikaylagawarecki) [inline test/test_scatter_gather_ops.py:88]: One more thing, could you please add some sample inputs to the sample_inputs functions for the opinfos of scatter/scatter_add/scatter_reduce ([e.g. here](https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L8328)) where src is smaller than index for the case where `not requires_grad` (since we aren't supporting gradients for this case)? 

This will run more comprehensive tests (e.g. `python test/test_ops.py -k scatter` includes a test that your proposed changes work on [noncontiguous tensors](https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L575))

File: test/test_scatter_gather_ops.py (hunk 1 above)
Comment (redmagnos) [reply test/test_scatter_gather_ops.py:104]: No, I don't think so, unless I misunderstand your question. The `s` there is the size in a given dimension, so if `s = index.size(d) == 1` for some dimension `d` then it won't generate a `src` tensor that has size less than that along that dimension (since afaik there's no such thing as size `0` in a 3-dimensional tensor and this would break stuff). `s == 1` can only happen on `dim`, though, since all other dimensions in that generator are `>= 10`, so that loop is guaranteed to make at least one of the `src` dimension sizes smaller than the corresponding `index` dimension size.

File: test/test_scatter_gather_ops.py (hunk 2 above)
Comment (redmagnos) [reply test/test_scatter_gather_ops.py:88]: Roger that!

Also, I am not sure these changes actually do work on noncontiguous tensors, I haven't tested them. I _think_ they should, if the logic works as I think it does, but I'll check.

File: torch/_tensor_docs.py [hunk 3]
```diff
@@ -3433,6 +3448,37 @@ def callable(a, b) -> number
...
+    >>> src = torch.arange(1, 13).reshape((2, 2, 3))
+    >>> src
+    tensor([[[ 1,  2,  3],
+             [ 4,  5,  6]],
+
+            [[ 7,  8,  9],
+             [10, 11, 12]]])
+    >>> gen = torch.Generator()
+    >>> gen.manual_seed(123456)
+    >>> index = torch.empty((2, 3, 4), dtype = torch.long).random_(0, 3, generator = gen)
```
[the remark this answers is not in the document]
Comment (redmagnos) [reply torch/_tensor_docs.py:3461]: Fair enough! I used the generator there because I wanted the index to be reasonably big (24 elements in that case) and I thought it'd look clunkier to write all 24 elements out by hand.

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu [hunk 4]
```diff
@@ -154,7 +154,85 @@ struct _cuda_scatter_gather_internal_kernel {
...
+    }
+
+    char* self_ptr = (char*)iter.data_ptr(0);
+    char* src_ptr = (char*)iter.data_ptr(1);
+    char* index_ptr = (char*)iter.data_ptr(2);
+
+    // We send the relevant index and source metadata
+    // to the device
+    int ndim = index_shape.size();
+    int64_t *index_shape_device, *index_strides_device, *src_shape_device, *src_strides_device;
```
[the remark this answers is not in the document]
Comment (redmagnos) [reply aten/src/ATen/native/cuda/ScatterGatherKernel.cu:192]: That's exactly the sort of thing I expected I'd miss/not know about, thank you! I'll look into changing that.

File: torch/_tensor_docs.py [hunk 5]
```diff
@@ -3433,6 +3448,37 @@ def callable(a, b) -> number
     tensor([[2.0000, 2.0000, 3.2300, 2.0000],
             [2.0000, 2.0000, 2.0000, 3.2300]])
 
+    >>> # Testing the case where index.size(d) > src.size(d)
```
[the remark this answers is not in the document]
Comment (redmagnos) [reply torch/_tensor_docs.py:3451]: Roger that

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp [hunk 6]
```diff
@@ -128,6 +128,54 @@ struct _cpu_scatter_gather_dim_loop {
...
+      int64_t idx_dim = index_data[i * index_dim_stride];
+      // we are not putting idx_dim in the error message because it disables
+      // loop optimization in clang-7
+      TORCH_CHECK(idx_dim >= 0 && idx_dim < index_upper_bound,
+        "index ", index_data[i * index_dim_stride],
+        " is out of bounds for dimension ", dim,
+        " with size ", index_upper_bound
+      );
+
+      // calculate the equivalent indices in the source tensor based on
```
[the remark this answers is not in the document]
Comment (redmagnos) [reply aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:157]: As I understand it, from that comment, if we restrided the tensors in that way the entries of `index` that go beyond those in `src` would be ignored rather than looping back, which I think was the desired behaviour? Unless I'm missing something.

File: test/test_scatter_gather_ops.py (hunk 1 above)
Comment (mikaylagawarecki) [reply test/test_scatter_gather_ops.py:104]: My bad, gotcha thanks!

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (redmagnos) [reply aten/src/ATen/native/cuda/ScatterGatherKernel.cu:192]: Okay, so, looking into it, I am not quite sure which of the various caching allocators I should use. There seem to be several options, the most likely-looking one being the `CUDACachingAllocator` one, but looking through examples in the codebase to check for usage does not make it immediately obvious how to then copy the data from host to device. I don't suppose you'd know of some resource or guide or particularly enlightening snippet that could point at what the best practices are?

File: test/test_scatter_gather_ops.py (hunk 2 above)
Comment (redmagnos) [reply test/test_scatter_gather_ops.py:88]: Testing manually with some small tensors the logic does seem to work for noncontiguous tensors, and I've added sample inputs that I _think_ do the thing, if I understood them correctly.

**Comment by redmagnos** [comment]
So I've been scratching my head about the Mysterious Failures in various tests here and finally saw the message the bot created a while ago talking about how I should be rebasing onto `viable/strict` rather than `master`, which might explain at least some of it? If so, whoops. I couldn't find this mentioned in the contributing docs but I might've missed it, and I definitely missed it the first time the bot commented here.

If that's the main reason for (at least some of) these failures then probably when I rebase onto `viable/strict` most of those failures should disappear. Hopefully!

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:157]: Right, only indices in the "intersection" are considered.

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/cuda/ScatterGatherKernel.cu:192]: Why not allocating a single buffer with a Tensor? You can get it's head with `t.data_ptr<type>()`.

File: test/test_scatter_gather_ops.py (hunk 2 above)
Comment (redmagnos) [reply test/test_scatter_gather_ops.py:88]: Okay, I had _not_ understood it correctly actually and my logic won't work for it. I'll think about how to fix it.

**Comment by janeyx99** [comment]
@redmagnos I suspect you may have rebased onto your fork's viable/strict or an outdated viable/strict, which is probably behind the current viable/strict in pytorch/pytorch. To rebase onto our viable/strict next time, I recommend doing something like:
```
git fetch --all    # to pull the most recent viable/strict
git rebase upstream/viable/strict OR git rebase origin/viable/strict     # you can check your .git/config to see whether upstream or origin points to pytorch/pytorch
```

In this case, I would recommend using `git cherry-pick` to select the commits you are trying to push onto a fresh branch. If you want to base that onto pytorch/pytorch's viable/strict, you can do something similar to:
```
git fetch --all
git checkout upstream/viable/strict 
git cherry-pick <commit-a> <commit-b> ...
git checkout -b less-strict-scatter-size-check
...
```
You can overwrite this current branch but you would have to manually remove all the additional reviewers.

Oh, another note for rebasing--you can use our rebase bot to rebase onto viable/strict by commenting `@pytorchbot rebase -s`. https://github.com/pytorch/pytorch/wiki/Bot-commands

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (redmagnos) [reply aten/src/ATen/native/cuda/ScatterGatherKernel.cu:192]: I'm not sure what you mean by that. What does seem necessary to me is a way to have an ordered list of strides and sizes for index and source, given that there's (as far as I could tell) no way to recover the information of what index a given offset is equivalent to in the original tensor with just the information otherwise provided to the loop. But there might be a way to access those values from the iterator itself or something in there that I missed, in which case passing it to device memory the way I did would be entirely unnecessary.

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:157]: Hey @redmagnos, this is a good point! While the intersection is one possible interpretation that would be simpler to implement (which I believe is why the issue was labelled as a good first issue), you are right that the (much more technically complex) wraparound behavior that you have implemented seems to be what #63265 was suggesting judging by the code block under alternatives. 

Since the wraparound behavior doesn't seem to be working for noncontiguous tensors at the moment (per your comment below) perhaps we could hold off on fixing this for a bit while we determine which interpretation of the enhancement would best serve our users.

If you had a use-case for this wraparound behavior that motivated you to take up this issue, it would be great if you could share as well :)

[the remark this answers is not in the document]
Comment (redmagnos) [reply aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:157]: I don't, actually; I assumed the original poster themself would have a reason, but the reason I picked this was more "it was marked as a good first issue and no one else seemed to be actively working on it".

With respect to noncontiguous tensors, I implemented something that seems to work for them and had been testing them earlier today, so since the work is already done I'll push it and you guys can take a look. And in case the "intersection" solution is the preferred one after all, it should be reasonably easy to pivot in that direction (since it is in fact just a matter of restriding with no change in the logic elsewhere at all).

File: test/test_scatter_gather_ops.py (hunk 2 above)
Comment (redmagnos) [reply test/test_scatter_gather_ops.py:88]: @mikaylagawarecki implemented the logic for CUDA, but as @nikitaved said we might want to go with the "intersection" interpretation of the request instead of the "wraparound" one, so I'll hold off on doing it for CPU until that's decided!

**Comment by Zettelkasten** [comment]
Hey, is there still development on this? I would love to see this merged.

**Comment by redmagnos** [comment]
@Zettelkasten It's still (as I understand it) undecided whether the desired behaviour is the logic I implemented (looping behaviour) or a simpler logic (overlapping behaviour) and until then this is on pause.
